File: src/main/js/codelity/lesson9.ts
function maxProfit(A: number[]): number {
    let maxProfit = 0;
    let min = 9999999999;
    for (const stockValue of A) {

        let profit = stockValue - min;

        if (profit > maxProfit) {
            maxProfit = stockValue - min;
        }

        if (stockValue < min) {
            min = stockValue;
        }

    }

    return maxProfit;
}

console.log(maxProfit([1, 2, 3, 4, 5]));


function MaxSliceSum(A: number[]) {
    let maxSliceSum = A[0];
    let sliceSum = maxSliceSum;

    for (let i = 1; i < A.length; i++) {
        const number = A[i];
        maxSliceSum = Math.max(sliceSum + number, number, maxSliceSum);
        if (sliceSum + number > 0) {
            if (number > sliceSum + number) {
                sliceSum = number
            } else {
                sliceSum += number;
            }
        } else {
            sliceSum = number;
        }
    }

    return maxSliceSum;


}

console.log(MaxSliceSum([3, -2, 3]));

File: src/main/java/codelity/Lesson9.java
package codelity;

public class Lesson9 {
    public int MaxProfit(int[] A) {
        int maxProfit = 0;
        int previousValue = Integer.MAX_VALUE;
        for (int stockValue : A) {
            int profit = stockValue - previousValue;

            if (profit > maxProfit) {
                maxProfit = stockValue - previousValue;
            }

            if (stockValue < previousValue) {
                previousValue = stockValue;
            }

        }

        return maxProfit;
    }

    public int MaxSliceSum(int[] A) {
        if (A.length == 0) {
            return 0;
        }

        int sum = A[0];
        int maxSum = sum;

        for (int i = 1; i < A.length; i++) {
            int number = A[i];

            sum = Integer.max(sum + number, number);
            maxSum = Integer.max( maxSum, sum);
        }

        return maxSum;
    }
}
